src/signatures.ts

```diff
@@ -0,0 +1,43 @@
+import {Signature, arrayify, splitSignature} from 'ethers/utils';
+import {State, hashState} from './contract/state';
+import {ethers} from 'ethers';
+import {SignedState} from '.';
+import {getChannelId} from './contract/channel';
+import {hashChallengeMessage} from './contract/challenge';
+import {toHex} from './hex-utils';
+
+export function getStateSignerAddress(signedState: SignedState): string {
+  const stateHash = hashState(signedState.state);
+  const recoveredAddress = ethers.utils.verifyMessage(
+    ethers.utils.arrayify(stateHash),
+    signedState.signature,
+  );
+  const {channel} = signedState.state;
+  const {participants} = channel;
+
+  if (participants.indexOf(recoveredAddress) < 0) {
+    throw new Error(
+      `Recovered address ${recoveredAddress} is not a participant in channel ${getChannelId(
+        channel,
+      )}`,
+    );
+  }
+  return recoveredAddress;
+}
+
+export async function signChallengeMessage(
+  signedStates: SignedState[],
+  privateKey: string,
+): Promise<Signature> {
+  const largestTurnNum = toHex(
+    signedStates.reduce(
+      (s1, s2) => (s1.state.turnNum >= s2.state.turnNum ? s1 : s2),
```

Comment:
Wondering if we want `>` here instead of `>=`, so that in the event of repeated `turnNum`s we default to the one that appears later in the array?

Reply:
Since we're just returning the `turnNum` I don't think it matters which one is returned, but in other cases where we use the whole `state` that makes sense.

Reply:
(Note that `Math.max` returns "NaN" when one of its input is not a number.)